[MOD-136][Improvement] Loading for the dashboard

laurenbarker opened the pull request:
## TODO
- [x] Use the cos-fork of ember-content-placeholders

## Changes
* Install ember-concurrency
* Move loading data out of routes
* Add skeleton screens for dashboard

![dashboard-skeleton](https://user-images.githubusercontent.com/7131985/32458369-3658fb42-c2fa-11e7-8c02-c23878682493.gif)

laurenbarker committed:
Use tasks on dashboard

laurenbarker committed:
Use tasks on dashboard and action-feed

Use skeleton screen on action-feed

laurenbarker committed:
Use cos-forks version of emebr-content-placeholders

aaxelb commented on app/components/action-feed/template.hbs:
```diff
@@ -1,14 +1,18 @@
-{{#if listLoading}}
-    {{loading-indicator classes='ball-dark text-center loading'}}
+{{#if (and loadActions.isRunning (not loadingMore))}}
```
I think we could get rid of `loadingMore` and DRY this up a little...
```hbs
{{#each actionsList ...}}
  entry
{{else if (not loadActions.isRunning)}}
  No actions
{{/if}}
{{#if loadActions.isRunning}}
  dummy list
{{else if moreActions}}
  see more
{{/if}}
```

aaxelb commented on app/controllers/index.js:
```diff
@@ -35,4 +39,16 @@ export default Controller.extend({
+    },
+
+    fetchData: task(function* () {
+        yield timeout(1000);
```
Is this left over from debugging?

aaxelb commented on app/routes/index.js:
```diff
@@ -1,26 +1,4 @@
-import { inject as service } from '@ember/service';
 import Route from '@ember/routing/route';
 
 export default Route.extend({
```
Can we delete this file?

aaxelb commented on app/components/action-feed/component.js:
```diff
@@ -22,43 +23,41 @@ export default Component.extend({
+        const page = this.get('page');
+        try {
+            const results = yield this.get('currentUser.user')
+                .then(user => this.get('store').queryHasMany(user, 'actions', { page }));
```
This isn't directly related to your changes, but would it load faster if we added `embed: 'target'` here?

laurenbarker replied:
Embedding isn't faster but it prevents the piecemeal loading.

aaxelb commented on app/components/action-feed/component.js:
```diff
@@ -22,43 +23,41 @@ export default Component.extend({
+    loadActions: task(function* () {
+        const page = this.get('page');
+        try {
+            const results = yield this.get('currentUser.user')
```
Thoughts on the ember-concurrency idiom instead of chaining promises?
```js
const user = yield this.get('currentUser.user');
const actions = yield this.get('store').queryHasMany(user, ...);
this.get('actionsList').pushObjects(actions.toArray());
...
```

aaxelb commented on app/components/action-feed/component.js:
```diff
@@ -22,43 +23,41 @@ export default Component.extend({
         this.incrementProperty('page');
         this.set('loadingMore', true);
-        this.loadPage();
+        this.get('loadActions').perform();
```
If you add `.drop()` to `loadActions`, seems like we could safely increment `page` inside it and get rid of `nextPage`

aaxelb commented on app/components/action-feed-entry/style.scss:
```diff
@@ -1,10 +1,6 @@
   clear: both;
 
-  .entry-body {
-    padding: 20px 15px;
```
Why no padding?
![screen shot 2017-11-06 at 4 46 51 pm](https://user-images.githubusercontent.com/6776190/32465674-45776420-c312-11e7-81c7-752735b48d30.png)

laurenbarker replied:
Aw...I forgot about the hover box. The top of the list looks out of line with the provider box with top padding.

aaxelb reviewed (changes requested):
Overall, looks awesome, just a smattering of small things.

Tiny bug: I think something happened moving data out of the route -- when I refresh the settings page, the count on the moderation tab is initially correct, then changes to `0` after a second.

...and there were some tests merged earlier today that this breaks, sorry about that.

laurenbarker committed:
Fix action entry styling

Remove extra file

Chaining promises --> ember-concurrency

Embed target

DRY up code

laurenbarker committed:
Move index.js to a component so it doesn't interfere with other pages

laurenbarker committed:
Fix tests

laurenbarker commented:
I don't know why the build isn't running. It passed on my branch...

aaxelb commented on app/components/action-feed-entry-skeleton/template.hbs:
```diff
@@ -0,0 +1,11 @@
+        <div class="icon--skeleton">{{placeholder.icon}}</div>
+    {{/content-placeholders}}
+    <div class="action-body--skeleton">
+        {{#content-placeholders as |placeholder|}}
```
Could these two `{{#content-placeholders as |placeholder|}}` blocks be combined in one?

aaxelb commented on app/components/landing-page/component.js:
```diff
@@ -0,0 +1,55 @@
+            this.transitionToRoute('preprints.provider.preprint-detail', provider.get('id'), reviewable.get('id'));
+        },
+        setupProvider(provider) {
+            this.transitionToRoute('preprints.provider.setup', provider.id);
```
`Uncaught TypeError: this.transitionToRoute is not a function`

The `router` service would help.

aaxelb commented on app/components/landing-page/template.hbs:
```diff
@@ -0,0 +1,47 @@
+    {{loading-indicator-page}}
+{{else}}
+    {{#if showDashboard}}
+        <div class='reviews-dashboard content'>
```
If we're doing all components, I think it'd be nice to make the landing page and dashboard separate components. Maybe the logic to switch between them would make sense on the controller? Not super important, but now seems like a convenient time to clean this up, if you don't mind...

laurenbarker replied:
The request for providers is what messes up the other pages. So I can move some stuff around but I don't think the logic to switch between them can go in the controller since it's called when the app refreshes no matter what page the user is on (that was the bug on the settings page).

aaxelb commented on app/components/action-feed/component.js:
```diff
@@ -18,47 +19,43 @@ export default Component.extend({
-            totalPages: Math.ceil(response.get('links.meta.total') / response.get('links.meta.per_page')),
-            listLoading: false,
+            actionsList: [],
             loadingMore: false,
```
`loadingMore` is gone

aaxelb reviewed (changes requested):
Found a little bug and had a few suggestions.

laurenbarker committed:
Move landing page into separate components

laurenbarker committed:
Remove unused variables

Condense placeholders

aaxelb commented on app/components/reviews-dashboard/template.hbs:
```diff
@@ -0,0 +1,23 @@
+            </div>
+            <div class='col-md-8 col-md-pull-4'>
+                {{action-feed
+                    toDetail=(action transitionToDetail)
```
Since `transitionToDetail` is already a closure action, I think you could just do `toDetail=transitionToDetail`. But more than that, it feels wrong to pass this action so far down (controller => reviews-dashboard => action-feed => action-feed-entry) when action-feed-entry could just use the router service and trigger a transition all by itself.

laurenbarker replied:
I was trying to avoid that since it seems very anti-Ember but we are ignoring pretty much everything they recommend so I guess it doesn't really matter. I do hate passing things down 10 levels.

aaxelb replied:
I think it'd make more sense if these components were meant to be reusable outside this app, or the action varied depending on other state, or it did anything other than a simple transition.

They added `router`, though, so I figure we may as well use it.

laurenbarker replied:
@aaxelb, bad news. We can't use the router service yet because their is a [bug](https://github.com/emberjs/ember.js/pull/15613) that breaks when all query params aren't present. So when we try and use `this.get('router').transitionTo(...)` we always get an error about the [ticket query parameter on the application route](https://github.com/CenterForOpenScience/ember-osf/blob/develop/addon/mixins/osf-cookie-login-controller.js#L20).
```
"Assertion Failed: You passed the `false` query parameter during a transition into application, please update to ticket"
``` 

Fix release:
> Apologies for not getting this in for 2.16, I'll get the changes into the next 2.17 beta release and once some folks can confirm all is working well for us I'll pull down into 2.16 for the LTS.

aaxelb replied:
That fix is in [2.16.1](https://github.com/emberjs/ember.js/releases/tag/v2.16.1)... we should upgrade. In another ticket.

laurenbarker replied:
Oh nice, I missed that. I'll add a ticket 👍

aaxelb commented on app/components/landing-page/template.hbs:
```diff
@@ -0,0 +1,12 @@
+            <div class="text-center m-t-lg col-xs-12">
+                <div class="m-b-lg reviews-brand"></div>
+                {{#if showSetup}}
+                    {{start-moderating action=(action setupProvider) providers=providersToSetup}}
```
Same thing here: It seems cleaner for start-moderating to use the router service and trigger the transition itself, rather than passing `setupProvider` all the way down from the controller.

aaxelb reviewed (changes requested):
Just one last thing, if it makes sense to you.
